src/Orleans/Serialization/OrleansJsonSerializer.cs

```diff
@@ -237,5 +233,29 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
         }
     }
 
+    class GuidConverter : JsonConverter
+    {
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType.IsAssignableFrom(typeof(Guid));
+        }
+
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            Guid guid = (Guid)value;
+            writer.WriteStartObject();
+            writer.WritePropertyName("Guid");
+            writer.WriteValue(value.ToString());
+            writer.WriteEndObject();
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            JObject jo = JObject.Load(reader);
+            Guid guid = Guid.Parse(jo["Guid"].ToObject<string>());
+            return guid;
+        }
```

Comment:
The implementation of WriteJson/readJson - there is indeed like 10 different ways it could have been implemented. It is  not my goal to optimize the Json serializer (for Orleans or at all).
I followed up on that issue since it simply did not make sense and kept bothering me. I just could not believe there is no way to have proper Guid deserialization in Json.

And the Json.Net guys [was not helpful](https://github.com/JamesNK/Newtonsoft.Json/issues/762), but he did hint in the past that there is a way - https://github.com/JamesNK/Newtonsoft.Json/issues/728.